### ISSUE TITLE
PNS - A malicious user can block others from depositing into the vault

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-07-sense-points-marketplace-judging#101]
PNS

Medium

# A malicious user can block others from depositing into the vault

## Summary

A malicious user can block legitimate users from depositing tokens into a vault with a set limit by front-running their transactions. The attacker deposits tokens up to the limit, causing subsequent deposit transactions to fail due to exceeding the cap.

## Vulnerability Detail

The `PointTokenVault` contract allows immediate deposits and withdrawals of tokens. The admin can set a limit on the number of tokens that can be deposited into the vault. Here's the attack scenario:

1. The admin adds a token to the whitelist and sets a limit for its deposits.
2. A malicious user monitors the mempool for incoming deposit transactions.
3. The malicious user deposits an amount of tokens that causes the subsequent deposit transaction (the one observed in the mempool) to exceed the limit and revert.
4. The malicious user then withdraws their tokens.

This scenario would effectively block most deposit attempts, as legitimate transactions would repeatedly fail due to the limit being temporarily reached.

## Impact

This vulnerability could discourage users from using the protocol, leading to a reduction in deposits and, consequently, a loss of revenue from fees.

## Code Snippet

```solidity
File: point-tokenization-vault/contracts/PointTokenVault.sol
  113:     // Rebasing and fee-on-transfer tokens must be wrapped before depositing.
  114:     function deposit(ERC20 _token, uint256 _amount, address _receiver) public {
  115:         uint256 cap = caps[address(_token)]; //audit
  116: 
  117:         if (cap != type(uint256).max) {
  118:             if (_amount + _token.balanceOf(address(this)) > cap) {
  119:                 revert DepositExceedsCap();
  120:             }
  121:         }
  122: 
  123:         _token.safeTransferFrom(msg.sender, address(this), _amount);
  124: 
  125:         balances[_receiver][_token] += _amount;
  126: 
  127:         emit Deposit(msg.sender, _receiver, address(_token), _amount);
  128:     }
  129: 
  130:     function withdraw(ERC20 _token, uint256 _amount, address _receiver) public {
  131:         balances[msg.sender][_token] -= _amount;
  132: 
  133:         _token.safeTransfer(_receiver, _amount);
  134: 
  135:         emit Withdraw(msg.sender, _receiver, address(_token), _amount);
  136:     }
```

- [PointTokenVault.deposit:114](https://github.com/sherlock-audit/2024-07-sense-points-marketplace/blob/076bf833f4dc1418e93c8172e4a4110344f1c812/point-tokenization-vault/contracts/PointTokenVault.sol#L114-L114)
- [PointTokenVault.withdraw:130](https://github.com/sherlock-audit/2024-07-sense-points-marketplace/blob/076bf833f4dc1418e93c8172e4a4110344f1c812/point-tokenization-vault/contracts/PointTokenVault.sol#L130-L130)

## Tool used

Manual Review

## Recommendation

Several solutions could mitigate this issue, depending on the business requirements of the project:

1. Implement a limit on individual deposit amounts.
2. Introduce a fee for deposits and withdrawals to disincentivize rapid deposits and withdrawals.
3. Enforce a time lock on deposits to prevent immediate withdrawals.

In summary, the goal is to increase the cost of such an attack and make it financially unviable.
